exodus_lambda/functions/verify_source.py

```diff
@@ -0,0 +1,64 @@
+import os
+import subprocess
+
+import boto3
+
+
+def lambda_handler(event, context):
+    """Verifies the commit delivered to the pipeline by GitHub has a
+    tag signed by a trusted entity.
+
+    Receives the commit ID from the pipeline via UserParameters.
+
+    This function must be manually deployed prior to pipeline creation.
+    """
+    # pylint: disable=unused-argument
+
+    job_id = event["CodePipeline.job"]["id"]
+    job_data = event["CodePipeline.job"]["data"]
+    git_commit = job_data["actionConfiguration"]["configuration"][
+        "UserParameters"
+    ]
+
+    bucket_name = "exodus-pipeline-artifacts"
+    whitelist_name = ".whitelist.gpg"
+
+    s3_client = boto3.client("s3")
+    pipeline_client = boto3.client("codepipeline")
+
+    try:
+        # 1) Download and import whitelisted GPG keys
+        s3_client.download_file(bucket_name, whitelist_name)
```

Comment:
I think it may be better to drop this whitelist download/re-upload stuff and just embed the keys into the pipeline.  For example, included in the UserParameter?  YAML has good support for multi-line strings so it is not going to be too ugly compared to the .whitelist.gpg file.

Right now it seems to be like this:

- The pipeline is not self-updating, if you want to change the steps in the pipeline or their config, those will not be updated simply by making the change in "deploy" branch.  You need to explicitly arrange an update to the pipeline.
- An exception is made for the GPG key whitelist, we wrote extra code to retrieve and store the whitelist during deployment so that this part of the pipeline's config can be self-updating.

I have to wonder: why do we need that exception?  It's not like this will need to be updated frequently.

Thinking of it another way.  If I wanted to change, say, the git branch used by the pipeline or the mailing list used for notifications, the process would be: update the YAML file, then manually run some documented command(s) to redeploy the pipeline with the updated config.  Right?

Is there any need for the process for updates to the GPG key list to be any different?  Isn't it just another piece of config for the pipeline, like the repo URL/branch or mailing list, which could be updated in the same way.

Reply:
That makes sense. I'll put pub keys directly into the invoke action's UserParameters.